doc/man3/DEFINE_SPARSE_ARRAY_OF.pod

```diff
@@ -0,0 +1,111 @@
+sa_TYPE_num, sa_TYPE_doall, sa_TYPE_doall_arg, sa_TYPE_get, sa_TYPE_set
+- sparse array container
+
+=head1 SYNOPSIS
+
+=for comment generic
+
+ #include "internal/sparse_array.h"
+
+ typedef struct sparse_array_st OPENSSL_SA;
+
+ SPARSE_ARRAY_OF(TYPE)
+ DEFINE_SPARSE_ARRAY_OF(TYPE)
+
+ SPARSE_ARRAY_OF(TYPE) *sa_TYPE_new(void);
+ void sa_TYPE_free(const SPARSE_ARRAY_OF(TYPE) *sa);
+ void sa_TYPE_free_leaves(const SPARSE_ARRAY_OF(TYPE) *sa);
+ int sa_TYPE_num(const SPARSE_ARRAY_OF(TYPE) *sa);
+ void sa_TYPE_doall(OPENSSL_SA *sa, void (*leaf)(void *));
+ void sa_TYPE_doall_arg(OPENSSL_SA *sa, void (*leaf)(void *), void *arg);
+ TYPE *sa_TYPE_get(const SPARSE_ARRAY_OF(TYPE) *sa, size_t idx);
+ int sa_TYPE_set(const SPARSE_ARRAY_OF(TYPE) *sa, size_t idx, TYPE *value);
+
+=head1 DESCRIPTION
+
+SPARSE_ARRAY_OF() returns the name for a sparse array of the specified
+B<TYPE>.  DEFINE_STACK_OF() creates set of functions for a sparse array
+of B<TYPE>. This will mean that type B<TYPE> is stored in each sparse
+array, the type is referenced by SPARSE_ARRAY_OF(TYPE) and each function
+name begins with I<sa_TYPE_>. For example:
+
+ TYPE *sa_TYPE_get(SPARSE_ARRAY_OF(TYPE) *sa, size_t idx);
+
+sa_TYPE_num() returns the number of elements in B<sa> or 0 if B<sa> is
+B<NULL>.
+
+sa_TYPE_get() returns element B<idx> in B<sa>, where B<idx> starts at
+zero. If B<idx> refers to a value that has not been set then B<NULL> is
+returned.
+
+sa_TYPE_set() sets element B<idx> in B<sa> to B<value>, where B<idx> starts at
+zero. The sparse array will be resized as required.
+
+sa_TYPE_new() allocates a new empty sparse array.
+
+sa_TYPE_free() frees up the B<sa> structure. It does B<not> free up any
+elements of B<sa>. After this call B<sa> is no longer valid.
+
+sa_TYPE_free_leaves() frees up the B<sa> structure and all of its elements.
+After this call B<sa> is no longer valid.
+
+sa_TYPE_doall() calls the function B<leaf> for each element in B<sa> in
+ascending index order.
+
+sa_TYPE_doall() calls the function B<leaf> for each element in B<sa> in
+ascending index order. The argument B<arg> is passed to each call of B<leaf>.
+
+=head1 NOTES
+
+Sparse arrays are an internal data structure and should B<not> be used by user
+applications.
+
+Care should be taken when accessing sparse arrays in multi-threaded environments.
+The sa_TYPE_set operation can cause the internal structure of the sparse array
+to change whichs causes race conditions if the sparse array is accessed in
+a different thread.
+
+SPARSE_ARRAY_OF() and DEFINE_SPARSE_ARRAY_OF() are implemented as macros.
+
+The underlying utility B<OPENSSL_SA_> API should not be used directly.
+It defines these functions: OPENSSL_SA_doall, OPENSSL_SA_doall_arg,
+OPENSSL_SA_free, OPENSSL_SA_free_leaves, OPENSSL_SA_get, OPENSSL_SA_new,
+OPENSSL_SA_num and OPENSSL_SA_set.
+
+=head1 RETURN VALUES
+
+sk_TYPE_num() returns the number of elements in the sparse array or B<0> if the
```

Comment:
Is the `sk_` prefix intentional here and in the following names?

Reply:
Nope, finger memory.